### ISSUE TITLE
empty catch clause in observable.trigger()

### ISSUE THREAD

**Issue by yuretz** [riot/observable#10]
The empty catch clause [here](https://github.com/riot/observable/blob/master/lib/index.js#L106) is really not a good idea. I can understand that the intent probably was to not let exceptions stop the event propagation and ensure the observable internal state consistency, but just hiding those exception "under the carpet" like this is a commonly considered to be a bad practice.

What I propose is to either change it to try..finally, reset the state in finally block, and let exceptions to be thrown, or, as an alternative, introduce a possibility of adding some kind of user-defined error-handling callback. This could probably be even done through observable events with some predefined name, like
`observable.on('error', function(error) {/* some handling code here */})`

What do you think?

**Comment by GianlucaGuarini** [comment]
@yuretz thank you now the errors can be easily caught

**Comment by yuretz** [comment]
Cool! Thanks.

**Comment by mh-cbon** [comment]
Hi,

using riot/2.3.15 i m facing problem with voided errors within event handlers.

My handler will fire once and no errors will be produced anywhere.

I tried to bind error event, but it did not report anything.

What can i do ? Should I wait ?

---

edit

In the test here

``` js
 it('Do not block callback throwing errors', function() {

    el.on('error', function() { counter++ })
    el.on('event', function() { counter++; throw 'OH NOES!' })
    el.on('event', function() { counter++ })

    expect(el.trigger).withArgs('event').to.throwException()
    expect(counter).to.be(1)
  })
```

I m unsure what it is expected but if the errored handlers must not block then, i guess, counter should eq 2.
Also i d used two variables to count `error` and `event` triggers and make this test more clear about its expectations.
mho, hth.

**Comment by GianlucaGuarini** [comment]
@mh-cbon this feature was removed in riot 2.3.13 http://riotjs.com/release-notes/#january-11-2016 and it was a terrible mistake adding it to riot

**Comment by mh-cbon** [comment]
@GianlucaGuarini 

thanks for pointing information, I read it too.

But still, i do have to write the event handler in this fashion,

``` js
  var vlc       = riot.observable()
// ...
      opts.vlc && opts.vlc.on('status', function (s) {
        try{
          status = s.state;
          if (status==='stopped') {
            timeHandle.value = 0
            timeHandle.disabled = 'disabled'

          } else if (status==='playing') {
            timeHandle.disabled = null;
            timeHandle.value = parseInt(s.position*100);

          } else if (status==='paused') {
            timeHandle.disabled = null;
            timeHandle.value = parseInt(s.position*100);

          } else {
            console.log(status)
          }
        }catch(ex){
          console.log(ex)
        }
      })
```

Which is not so cool for such a good lib until now.

I must also mention that binding error event do not help me at all.

I double check my observable dependency here, see what i got,

``` json
{
  "_args": [
    [
      "riot-observable@^2.3.14",
      "/home/mh-cbon/projects/vrc/node_modules/riot"
    ]
  ],
  "_from": "riot-observable@>=2.3.14 <3.0.0",
  "_id": "riot-observable@2.3.14",
  "_inCache": true,
  "_location": "/riot-observable",
  "_nodeVersion": "5.1.1",
  "_npmUser": {
    "email": "gianluca.guarini@gmail.com",
    "name": "gianlucaguarini"
  },
  "_npmVersion": "3.5.0",
  "_phantomChildren": {},
  "_requested": {
    "name": "riot-observable",
    "raw": "riot-observable@^2.3.14",
    "rawSpec": "^2.3.14",
    "scope": null,
    "spec": ">=2.3.14 <3.0.0",
    "type": "range"
  },
  "_requiredBy": [
    "/riot",
    "/riot-route"
  ],
  "_resolved": "https://registry.npmjs.org/riot-observable/-/riot-observable-2.3.14.tgz",
  "_shasum": "97e3b0c0372a61a6f1a26d5f72742a23f9dafa97",
  "_shrinkwrap": null,
  "_spec": "riot-observable@^2.3.14",
  "_where": "/home/mh-cbon/projects/vrc/node_modules/riot",
  "author": {
    "name": "Muut, Inc. and other contributors"
  },
  "bugs": {
    "url": "https://github.com/riot/observable/issues"
  },
  "dependencies": {},
  "description": "Simple script to send and receive events",
  "devDependencies": {
    "coveralls": "^2.11.6",
    "eslint": "^1.10.3",
    "expect.js": "^0.3.1",
    "karma": "^0.13.15",
    "karma-coverage": "^0.5.3",
    "karma-mocha": "^0.2.1",
    "karma-phantomjs-launcher": "^0.2.1",
    "mocha": "^2.3.4",
    "phantomjs": "^1.9.19"
  },
  "directories": {
    "lib": "lib"
  },
  "dist": {
    "shasum": "97e3b0c0372a61a6f1a26d5f72742a23f9dafa97",
    "tarball": "http://registry.npmjs.org/riot-observable/-/riot-observable-2.3.14.tgz"
  },
  "gitHead": "1aa6425cf9ba7b756491e8348a9b8ce43eacf1a7",
  "homepage": "https://github.com/riot/observable#readme",
  "installable": true,
  "jsnext:main": "dist/es6.observable.js",
  "keywords": [
    "events",
    "observable",
    "riot"
  ],
  "license": "MIT",
  "main": "dist/observable.js",
  "maintainers": [
    {
      "name": "gianlucaguarini",
      "email": "gianluca.guarini@gmail.com"
    },
    {
      "name": "tipiirai",
      "email": "tero@muut.com"
    },
    {
      "name": "cognitom",
      "email": "kawamura@cognitom.com"
    }
  ],
  "name": "riot-observable",
  "optionalDependencies": {},
  "repository": {
    "type": "git",
    "url": "git+https://github.com/riot/observable.git"
  },
  "scripts": {
    "prepublish": "make && make test",
    "test": "make test"
  },
  "version": "2.3.14"
}
```

My dep indicates **2.3.14**, but your repo indicate **2.3.13**, https://github.com/riot/observable/blob/master/package.json. 

sorry for that in advance but i have to ask as straight as it come, 
wtf ? Is it relevant ?

**Comment by GianlucaGuarini** [comment]
@mh-cbon thanks for raising this up I will fix the package.json in riot-observable. But on npm the latest version available is `riot-observable@2.3.14`

**Comment by mh-cbon** [comment]
Thanks for version, indeed it is fixed and probably not related. May i suggest https://docs.npmjs.com/cli/version ?

Also i gone though https://github.com/riot/riot/issues/1483

Maybe I should mention my error, 
I was calling a wrong `conole.log` and could not figure out it becasue my error was not sent to any output.

I may also mention that the error handler would be called once only.

It was similar to this,

``` js
      opts.vlc && opts.vlc.on('status', function (s) {
        console.log('ok')
          conole.log('r')
          status = s.state;
          if (status==='stopped') {
            timeHandle.value = 0
            timeHandle.disabled = 'disabled'

          } else if (status==='playing') {
            timeHandle.disabled = null;
            timeHandle.value = Math.parseInt(s.position*100);

          } else if (status==='paused') {
            timeHandle.disabled = null;
            timeHandle.value = Math.parseInt(s.position*100);

          } else {
            console.log(status)
          }
      })
    })
```

hth
